### PR TITLE
Add `extract_sequences`, `rfft` and `stft` to ops.math

### LINKED ISSUE #697: Add `stft` to keras_core.ops
Hi keras-core team

I'm trying to use keras-core to build an audio model. However, I've found that there are certain operations that are missing for audio processing
- `tf.signal.frame`
- `tf.signal.rfft`
- `tf.signal.stft`
etc...

I have already implemented `stft` for all backends. Is it good to add to keras-core?
If that's the case, would it be better to separate `rfft` and `frame` from `stft`?

The working example with the the results that are passed with `np.testing.assert_allclose`
(I have implemented `stft` with the signature from librosa because I think it is a more mature library for audio processing)

<details>

```python
import math

import jax
import librosa as lr
import numpy as np
import scipy.signal
import torch
import tensorflow as tf

from keras_core.backend.jax.core import (
    convert_to_tensor as jax_convert_to_tensor,
)
from keras_core.backend.numpy.core import (
    convert_to_tensor as np_convert_to_tensor,
)
from keras_core.backend.tensorflow.core import (
    convert_to_tensor as tf_convert_to_tensor,
)
from keras_core.backend.torch.core import (
    convert_to_tensor as torch_convert_to_tensor,
)


# librosa
def librosa_stft(signal, n_fft, hop_length, win_length, center, window="hann"):
    return lr.stft(
        y=signal,
        n_fft=n_fft,
        hop_length=hop_length,
        win_length=win_length,
        window=window,
        center=center,
    )


# tensorflow
def tf_stft(signal, n_fft, hop_length, win_length, center, window="hann"):
    signal = tf_convert_to_tensor(signal)
    rank = tf.rank(signal)
    if center:
        if rank == 1:
            padding = [[n_fft // 2, n_fft // 2]]
        else:
            padding = [[0, 0], [n_fft // 2, n_fft // 2]]
        signal = tf.pad(signal, padding, mode="REFLECT")
    if isinstance(window, str):
        if window == "hamming":
            win = tf.signal.hamming_window(win_length, periodic=True)
        elif window == "hann":
            win = tf.signal.hann_window(win_length, periodic=True)
        else:
            raise NotImplementedError()
    else:
        win = tf_convert_to_tensor(window, dtype=signal.dtype)
    left_pad = (n_fft - win_length) // 2
    right_pad = n_fft - win_length - left_pad
    win = tf.pad(win, [[left_pad, right_pad]])
    framed_signals = tf.signal.frame(
        signal, frame_length=n_fft, frame_step=hop_length
    )
    framed_signals *= win
    framed_signals = tf.signal.rfft(framed_signals, [n_fft])
    if rank == 1:
        framed_signals = tf.transpose(framed_signals, perm=[1, 0])
    else:
        framed_signals = tf.transpose(framed_signals, perm=[0, 2, 1])
    return framed_signals


# torch
def torch_stft(signal, n_fft, hop_length, win_length, center, window="hann"):
    signal = torch_convert_to_tensor(signal)
    if isinstance(window, str):
        if window == "hamming":
            win = torch.hamming_window(win_length, periodic=True)
        elif window == "hann":
            win = torch.hann_window(win_length, periodic=True)
        else:
            raise NotImplementedError()
    else:
        win = torch_convert_to_tensor(window)
    return torch.stft(
        signal,
        n_fft=n_fft,
        hop_length=hop_length,
        win_length=win_length,
        window=win,
        center=center,
        return_complex=True,
    )


# jax
def jax_stft(signal, n_fft, hop_length, win_length, center, window="hann"):
    signal = jax_convert_to_tensor(signal)
    rank = len(signal.shape)
    if center:
        padding = [(0, 0) for _ in range(rank)]
        padding[-1] = (n_fft // 2, n_fft // 2)
        signal = jax.numpy.pad(signal, padding, mode="reflect")
    if isinstance(window, str):
        win = jax.numpy.array(
            scipy.signal.get_window(window, win_length), dtype=signal.dtype
        )
    else:
        win = jax_convert_to_tensor(window)
    left_pad = (n_fft - win_length) // 2
    right_pad = n_fft - win_length - left_pad
    win = jax.numpy.pad(win, [[left_pad, right_pad]])

    # equivalent to tf.signal.frame
    # https://gist.github.com/sourabh2k15/80adbf1c5861e727f7698fd66e51be39#file-collect_trace-py-L226
    *batch_shape, signal_length = signal.shape
    batch_shape = list(batch_shape)
    signal = jax.numpy.reshape(
        signal, (math.prod(batch_shape), signal_length, 1)
    )
    framed_signals = jax.lax.conv_general_dilated_patches(
        signal,
        (n_fft,),
        (hop_length,),
        "VALID",
        dimension_numbers=("NTC", "OIT", "NTC"),
    )
    framed_signals = jax.numpy.reshape(
        framed_signals, (*batch_shape, *framed_signals.shape[-2:])
    )
    framed_signals *= win
    framed_signals = jax.numpy.fft.rfft(framed_signals, n_fft)
    if rank == 1:
        framed_signals = jax.numpy.transpose(framed_signals, axes=[1, 0])
    else:
        framed_signals = jax.numpy.transpose(framed_signals, axes=[0, 2, 1])
    return framed_signals


# numpy
def numpy_stft(signal, n_fft, hop_length, win_length, center, window="hann"):
    signal = np_convert_to_tensor(signal)
    rank = len(signal.shape)
    if center:
        padding = [(0, 0) for _ in range(rank)]
        padding[-1] = (n_fft // 2, n_fft // 2)
        signal = np.pad(signal, padding, mode="reflect")
    if isinstance(window, str):
        win = np.array(
            scipy.signal.get_window(window, win_length), dtype=signal.dtype
        )
    else:
        win = np_convert_to_tensor(window)
    left_pad = (n_fft - win_length) // 2
    right_pad = n_fft - win_length - left_pad
    win = np.pad(win, [[left_pad, right_pad]])

    # equivalent to tf.signal.frame
    # https://github.com/scipy/scipy/blob/v1.11.1/scipy/signal/_spectral_py.py#L1928
    *batch_shape, _ = signal.shape
    batch_shape = list(batch_shape)
    shape = signal.shape[:-1] + (
        (signal.shape[-1] - (n_fft - hop_length)) // hop_length,
        n_fft,
    )
    strides = signal.strides[:-1] + (
        hop_length * signal.strides[-1],
        signal.strides[-1],
    )
    framed_signals = np.lib.stride_tricks.as_strided(
        signal, shape=shape, strides=strides
    )
    framed_signals = np.reshape(
        framed_signals, (*batch_shape, *framed_signals.shape[-2:])
    )
    framed_signals = np.multiply(framed_signals, win)
    framed_signals = np.fft.rfft(framed_signals, n_fft)
    if rank == 1:
        framed_signals = np.transpose(framed_signals, axes=[1, 0])
    else:
        framed_signals = np.transpose(framed_signals, axes=[0, 2, 1])
    return framed_signals


waveform, samplerate = lr.load(
    lr.util.example("brahms", hq=True), duration=10.0
)
for n_fft, hop_length, win_length in [
    (2048, 256, 1024),
    (1024, 128, 512),
    (123, 50, 80),
]:
    for center in [True, False]:
        for window_name in ["hann", "hamming"]:
            window = scipy.signal.get_window(window_name, win_length)
            window = window.astype(waveform.dtype)
            args = (waveform, n_fft, hop_length, win_length, center, window)

            # run stft with different backends
            lr_out = librosa_stft(*args)
            tf_out = tf_stft(*args).numpy()
            torch_out = torch_stft(*args).numpy()
            jax_out = np.asarray(jax_stft(*args))
            np_out = numpy_stft(*args)

            # np.testing.assert_allclose(np_out, lr_out, rtol=1e-5, atol=1e-5)
            np.testing.assert_allclose(np_out, tf_out, rtol=1e-5, atol=1e-5)
            np.testing.assert_allclose(np_out, torch_out, rtol=1e-5, atol=1e-5)
            np.testing.assert_allclose(np_out, jax_out, rtol=1e-5, atol=1e-5)
            print(
                f"Passed! n_fft={n_fft}, hop_length={hop_length}, "
                f"win_length={win_length}, center={center}, "
                f"window={window_name}"
            )
```
</details>

### PR DESCRIPTION
**Pull request by james77777778**
Fixes #697 

This PR adds `frame`, `rfft` and `stft` to ops.math with the corresponding docstrings and unit tests
The CPU version of tf needs `atol=1e-5, rtol=1e-5` to pass `rfft` and `stft` tests
(note: the GPU version is good with the default `1e-6`)

The signatures are following:
```python
def frame(x, frame_length, frame_step):
    ...

def rfft(x, fft_length=None):
    ...

# a mixture of tf and librosa/torch
def stft(x, frame_length, frame_step, fft_length, window="hann", center=True):
    ...
```

The visualization of STFT using keras-core:

```python
import librosa as lr
import matplotlib.pyplot as plt
import numpy as np

from keras_core.ops.math import stft

# load waveform from librosa
waveform, samplerate = lr.load(lr.util.example("trumpet"))
# stft using keras_core
real_out, imag_out = stft(
    waveform, frame_length=2048, frame_step=512, fft_length=2048
)
# visualize with librosa
real_out = np.transpose(real_out)
s_db = lr.amplitude_to_db(np.abs(real_out), ref=np.max)
fig, ax = plt.subplots()
img = lr.display.specshow(s_db, x_axis="time", y_axis="log", ax=ax)
ax.set(title='Using a logarithmic frequency axis')
fig.colorbar(img, ax=ax, format="%+2.f dB")
plt.savefig("output.png")
```

output

The `output.png` is almost the same as the example shown in librosa's doc
https://librosa.org/doc/main/auto_examples/plot_display.html#changing-axis-scales

### CODE REVIEW

**Comment by james77777778** [comment]
Furthermore, we can include following ops in ops.math:
- `ifft`
- `ifft2`
- `irfft`
- `rfft2` / `irfft2`
- `istft`

These might be considered as a follow-up?

File: keras_core/ops/math.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 """Commonly used math operations not included in NumPy."""
 
+import scipy.signal
```
Comment (fchollet) [inline keras_core/ops/math.py:3]: We should not depend on scipy if at all possible.

File: keras_core/ops/math.py [hunk 2]
```diff
@@ -450,3 +509,193 @@ def fft2(a):
...
+        )
+
+    def call(self, x):
+        return backend.math.rfft(x, n=self.n)
+
+
+@keras_core_export("keras_core.ops.rfft")
+def rfft(x, n=None):
+    """Computes the real-valued fast Fourier transform along the last axis of
+    the input.
```
Comment (fchollet) [inline keras_core/ops/math.py:545]: The one-line summaries should fit on a single line.

File: keras_core/ops/math.py [hunk 3]
```diff
@@ -450,3 +509,193 @@ def fft2(a):
...
+
+    Returns:
+        A tuple containing two tensors - the real and imaginary parts of the
+        output.
+
+    Examples:
+
+    >>> x = keras_core.ops.convert_to_tensor([0.0, 1.0, 2.0, 3.0, 4.0])
+    >>> rfft(x)
+    (array([10. , -2.5, -2.5]), array([0.        , 3.4409548 , 0.81229924]))
```
Comment (fchollet) [inline keras_core/ops/math.py:561]: Trim whitespace

File: keras_core/ops/math.py [hunk 4]
```diff
@@ -450,3 +509,193 @@ def fft2(a):
...
+
+    Args:
+        x: Input tensor.
+        frame_length: An integer representing the frame length in samples.
+        frame_step: An integer representing the frame hop size in samples.
+        fft_length: An integer representing the size of the FFT to apply. If not
+            specified, uses the smallest power of 2 enclosing `frame_length`.
+        window: A string or the tensor of the window. If `window` is a string,
+            it is passed to `scipy.signal.get_window` to generate the window
+            values, which are DFT-even by default. See
```
Comment (fchollet) [inline keras_core/ops/math.py:667]: The docstring should be standalone and should not refer to other sources to explain what argument values are valid. They should be listed here.

File: keras_core/ops/math.py [hunk 5]
```diff
@@ -450,3 +509,193 @@ def fft2(a):
...
+            KerasTensor(shape=new_shape, dtype=x.dtype),
+            KerasTensor(shape=new_shape, dtype=x.dtype),
+        )
+
+    def call(self, x):
+        return backend.math.rfft(x, n=self.n)
+
+
+@keras_core_export("keras_core.ops.rfft")
+def rfft(x, n=None):
```
Comment (fchollet) [inline keras_core/ops/math.py:543]: Surely `n` should be something like `fft_length`? Do we have any equivalent argument anywhere else in the API today?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_core/ops/math.py (hunk 1 above)
Comment (james77777778) [reply keras_core/ops/math.py]: I have refactored the implementation of `get_window`

The biggest change involves transferring the logic of `stft` from `keras_core.ops.math` to `keras_core.backend.*.math`

- tf: `tf.signal.hann_window` and `tf.signal.hamming_window`
- torch: `torch.hann_window` and `torch.hamming_window`
- jax & numpy: `scipy.signal.get_window`

File: keras_core/ops/math.py (hunk 2 above)
Comment (james77777778) [reply keras_core/ops/math.py]: Fixed. I have enhanced the docstring for the 3 operators (taken from tf)

File: keras_core/ops/math.py (hunk 3 above)
Comment (james77777778) [reply keras_core/ops/math.py]: Done

File: keras_core/ops/math.py (hunk 4 above)
Comment (james77777778) [reply keras_core/ops/math.py]: I have updated the docstring to guide users to select either `"hann"` or `"hamming"`
- tf & torch: only provides hann and hamming window
- numpy & jax: https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.get_window.html

File: keras_core/ops/math.py (hunk 5 above)
Comment (james77777778) [reply keras_core/ops/math.py]: Fixed.
I used `n` due to the naming convention from numpy but I think `fft_length` should be more clearer.

File: keras_core/ops/math.py [hunk 6]
```diff
@@ -314,6 +314,66 @@ def qr(x, mode="reduced"):
...
+        return KerasTensor(shape=new_shape, dtype=x.dtype)
+
+    def call(self, x):
+        return backend.math.frame(
+            x, frame_length=self.frame_length, frame_step=self.frame_step
+        )
+
+
+@keras_core_export("keras_core.ops.frame")
+def frame(x, frame_length, frame_step):
```
Comment (fchollet) [inline keras_core/ops/math.py:345]: I'm not 100% sure we should export `frame` to the public API as an op (the function here + the `Frame` class). How important is it compared to the other too?

File: keras_core/backend/numpy/math.py [hunk 7]
```diff
@@ -1,7 +1,10 @@
 import numpy as np
+import scipy.signal
```
Comment (fchollet) [inline keras_core/backend/numpy/math.py:2]: When using the numpy backend, scipy might not be installed. Prefer using `keras.utils.module_utils` to create a lazily imported package.

File: keras_core/backend/jax/math.py [hunk 8]
```diff
@@ -1,5 +1,11 @@
+import math
+
 import jax
 import jax.numpy as jnp
+import scipy.signal
```
Comment (fchollet) [inline keras_core/backend/jax/math.py:5]: Same here

File: keras_core/ops/math.py (hunk 6 above)
Comment (james77777778) [reply keras_core/ops/math.py:345]: I have discovered that it is used to build LongFormer (citation> 2k) in huggingface implementation
https://github.com/huggingface/transformers/blob/eec5841e9f440c795fb9292d009675d97a14f983/src/transformers/models/longformer/modeling_tf_longformer.py#L1107

There are about 1.6k search results for `tf.signal.frame`
https://github.com/search?q=tf.signal.frame&type=code

I believe that exporting it holds value. However, I agree that we should find a more suitable place for it.

Refs:
- tf: `tf.signal`
- torch: `torch.Tensor.unfold`
- jax & numpy: no direct api

Is a new namespace required for it?

File: keras_core/backend/numpy/math.py (hunk 7 above)
Comment (james77777778) [reply keras_core/backend/numpy/math.py]: Added

File: keras_core/backend/jax/math.py (hunk 8 above)
Comment (james77777778) [reply keras_core/backend/jax/math.py]: Done

File: keras_core/ops/math.py (hunk 6 above)
Comment (fchollet) [reply keras_core/ops/math.py:345]: Thanks for the info. OK to export. It is like a 1D equivalent of `extract_patches`, right? In this case perhaps we can standardize the API on `extract_frames` or `extract_windows` (which is also more self-explanatory to a user who does not know what it does).

We might also want to standardize the terminology, e.g. `step` -> `stride`. We also need to decide whether to refer to frames as "frames", "windows", or "sequences". A very similar utility exists for tf.data (`timeseries_dataset_from_array`) which uses "sequence".

What do you think?

Comment (fchollet) [reply keras_core/ops/math.py:345]: Sounds good to me!

Comment (james77777778) [reply keras_core/ops/math.py]: Updated.

**Review by fchollet — APPROVED** [review]
LGTM. Thank you for the contribution!
